### ISSUE TITLE
[SharedContainer::splitBlock] Container is not Independent

### ISSUE THREAD

**Issue by axiqia** [Shark-ML/Shark#259]
`normalizing_train.train(normalizer_train, training_data.inputs())'

this should fix your issue. Please remember not to train a normalizer on the test set and use

`normalizedData_test = transformInputs(test_data, normalizer_train);`

_Originally posted by @Ulfgard in https://github.com/Shark-ML/Shark/issues/51#issuecomment-189819470_
```
RegressionDataset raw_data;
importCSV(raw_data, argv[1], FIRST_COLUMN, 1, ' ');
raw_data.shuffle();     
bool removleMean = true;
Normalizer<RealVector> normalizer;
NormalizeComponentsUnitVariance<RealVector> normalizingTrainer(removleMean);
normalizingTrainer.train(normalizer, raw_data.inputs());
RegressionDataset data = transformInputs(raw_data, normalizer);

 //Split the dataset into a training and a test dataset
RegressionDataset   dataTest = splitAtElement(data, static_cast<std::size_t>(0.8*data.numberOfElements()));
```
After I normalized the raw data, I got an runtime error when I splitted the data into dataTest.

```
terminate called after throwing an instance of 'shark::Exception'
  what():  [SharedContainer::splitBlock] Container is not Independent
[1]    1440 abort (core dumped)  ./ExampleProject  actfwd.model

```

**Comment by Ulfgard** [comment]
arrgh! This might actually be a bug- I think transformInputs just copies the label container from raw_data, therefore the labels are still shared with raw_data. 

could you try whether the following before splitting helps?

data.labels().makeIndependent() 

I have to ask myself whether the independence check is doing more harm than good here.

**Comment by axiqia** [comment]
It works :)

**Comment by axiqia** [comment]
Dear @Ulfgard ,
Is there a API to transform a  RealVector using a Normalizer?
e.g.
```
      //load scaler from the file
      ifstream ifs2(argv[3]);
      TextInArchive ia2(ifs2);
      Normalizer<RealVector> normalizer2;
      normalizer2.load(ia2, 0);
      ifs2.close();

      double param[] = { 
      241.28,1, 1, 0, 1, 0, 1, 1, 1, 14745600, 14745600, 14745600, 14745600, 1, 0 ,0 ,1, 1, 1, 14745600, 14745600, 14745600, 14745600, 1
      };  
      RealVector onetest(23);
      std::copy(param, param+23, onetest.begin());
      RealVector points = transform(onetest, normalizer2);    
```

I only find `shark::transform (Data< T > const &data, Functor f)` [here](http://image.diku.dk/shark/doxygen_pages/html/group__shark__globals.html) .
Thank you :)

**Comment by Ulfgard** [comment]
normalizer is a model, so

normalizer2(onetest)

it would also work for Data<T> as argument if you happen to have many data points

**Comment by axiqia** [comment]
> normalizer is a model, so
> 
> normalizer2(onetest)
> 
> it would also work for Data as argument if you happen to have many data points

So I can also use `normalizer2(data.inputs())` and the result is the same as` transformInputs(data.inputs())`

Do I understand correctly?

**Comment by Ulfgard** [comment]
Yes,

this will internally just call transform, so you are fine. This is pure convenience
________________________________

**Comment by Ulfgard** [comment]
exactly
